(#448) Clarify logic in task.c; Add unit-test to validate public interfaces.

gapisback opened the pull request:
This commit adds a bit of code cleanup to the task system module, task.c . No new functionality or change-in-logic is being done. Add new unit-test, task_system_test, to validate internal state of the task system for simple task creation / execution.

---
NOTE: to the reviewer. This change set came out of my code walk through to understand the task system and its innards.
This was done in context of the design rework to support execution of Splinter using allocated shared memory segments.
While at it, I constructed a small C-unit test to work through task system interfaces and to confirm, via tests, expected outputs.

Some of this testing will also be helpful when we "cut-over" to running tests using shared segments.

gapisback commented on src/task.c:
```diff
@@ -10,23 +10,29 @@
 int              hook_init_done = 0;
 static int       num_hooks      = 0;
-static task_hook hooks[MAX_HOOKS];
+// RESOLVE: At 8 bytes per task_hook{}, we are burning up 16K of memory. Only 2 hooks
```
@ajhconway - We should discuss and reconsider this. 

When we, eventually, move to a shared memory implementation, I suspect (expect) that this array will need to be allocated from shared memory segment. That will cost us 16K, whereas something like 64 bytes should suffice if we reduce MAX_HOOKS to 8.

gapisback commented on src/task.c:
```diff
@@ -10,23 +10,29 @@
 
 // forward declarations that aren't part of the public API of the task system
-platform_status
+static platform_status
```
Many changes in this file add `static` to functions / prototypes to document clearly that this is a static function.
Improves code readability.

gapisback commented on src/task.c:
```diff
@@ -83,12 +104,21 @@ task_get_max_tid(task_system *ts)
+ * Return the bitmasks of tasks active. Mainly intended as a testing hook.
+ */
+uint64
+task_active_tasks_mask(task_system *ts)
```
Added this extern accessor function, used in unit-tests for validating bitmask.

gapisback commented on src/task.c:
```diff
@@ -149,7 +196,7 @@ task_register_this_thread(task_system *ts, uint64 scratch_size)
 task_deregister_this_thread(task_system *ts)
 {
-   uint64 tid = platform_get_tid();
+   threadid tid = platform_get_tid();
```
Use the right typedef.

gapisback commented on src/task.c:
```diff
@@ -168,6 +215,17 @@ typedef struct {
+ * function invokes that function bracketted by calls to register / deregister
+ * the thread with Splinter's task system.
+ *
+ * RESOLVE: I see references to this, ... but who calls this?
```
Let's discuss .... this, and then there seem to be a few more functions, that are written but never called.

Is this vestigial code that we could / should get rid of now, to reduce confusion ... [???].

gapisback commented on src/task.c:
```diff
@@ -185,6 +243,11 @@ task_invoke_with_hooks(void *func_and_args)
 
+/*
+ * -----------------------------------------------------------------------------
+ * task_create_thread_with_hooks() - Comment ?
```
Help! Can you help me add a meaningful prolog?

gapisback commented on src/task.c:
```diff
@@ -215,6 +278,15 @@ task_create_thread_with_hooks(platform_thread       *thread,
+ *  required registration with Splinter.
+ *
+ * Currently, this is active mainly in tests. It's also used to create
+ * background tasks, a feature that is currently not enabled.
```
@ajhconway - RESOLVE: What's the history on background threads?

Can I elicit some documentation from you and put it down here, so future readers can benefit?

gapisback replied:
Thanks, @ajhconway - I will update the prolog accordingly.

Have opened issue #450 requesting that we have to work on re-enabling background threads support.

ajhconway commented on src/task.c:
```diff
@@ -8,25 +8,32 @@
-static int       num_hooks      = 0;
+int        hook_init_done = 0;
+static int num_hooks      = 0;
+// RESOLVE: At 8 bytes per task_hook{}, we are burning up 16K of memory. Only 2
+// hooks are currently being connected. What about reducing MAX_HOOKS to, say,
+// 8?
```
This does feel a little much. Maybe something still hard to imagine blowing out like 64?

ajhconway commented on src/task.c:
```diff
@@ -68,9 +84,15 @@ init_threadid(task_system *ts)
 
 out:
    debug_assert(tid != INVALID_TID);
+
+   // Sets thread-ID, tracked as thread-local storage.
```
This is only true on Linux. It could be implemented differently on another platform.

I would just rephrase this as "Sets thread-ID, for example tracked as thread-local storage."

gapisback replied:
Done.

ajhconway commented on src/task.c:
```diff
@@ -98,19 +129,30 @@ register_init_tid_hook(void)
+
+   uint64 bitmask_val = *tid_bitmask;
+
+   // Ensure that caller is only clearing for a thread that's in-use.
+   platform_assert(!(bitmask_val & (1ULL << tid)),
+                   "Thread [%lu] is expected to be in-use. Bitmap: 0x%lx",
+                   tid,
+                   bitmask_val);
```
This is good

ajhconway commented on tests/unit/task_system_test.c:
```diff
@@ -0,0 +1,349 @@
+ */
+CTEST2(task_system, test_one_thread)
+{
+   pthread_t     new_thread;
+   thread_config thread_cfg;
+
+   ZERO_STRUCT(thread_cfg);
+
+   threadid main_thread_idx = platform_get_tid();
+   ASSERT_EQUAL(main_thread_idx, 0, "main_thread_idx=%lu", main_thread_idx);
+
+   // Setup thread-specific struct, needed for validation in thread's worker fn
+   thread_cfg.tasks                  = data->tasks;
+   thread_cfg.exp_thread_idx         = 1; // Main thread is at index 0.
+   thread_cfg.active_threads_bitmask = task_active_tasks_mask(data->tasks);
+
+   int rc = pthread_create(&new_thread, NULL, exec_one_thread, &thread_cfg);
+   ASSERT_EQUAL(0, rc);
```
This should be using `platform_thread` and `task_thread_create`.

ajhconway commented on tests/unit/task_system_test.c:
```diff
@@ -0,0 +1,349 @@
+ */
+CTEST2(task_system, test_multiple_threads)
+{
+   pthread_t      new_thread;
+   thread_config  thread_cfg[MAX_THREADS];
+   thread_config *thread_cfgp = NULL;
+   int            tctr        = 0;
+   int            rc          = 0;
+
+   ZERO_ARRAY(thread_cfg);
+   platform_default_log(" Before threads start, task_get_max_tid() = %lu\n",
+                        task_get_max_tid(data->tasks));
+
+   // Start-up n-threads, record their expected thread-IDs, which will be
+   // validated by the thread's execution function below.
+   for (tctr = 1, thread_cfgp = &thread_cfg[tctr];
+        tctr < ARRAY_SIZE(thread_cfg);
+        tctr++, thread_cfgp++)
+   {
+      // These are independent of the new thread's creation.
+      thread_cfgp->tasks          = data->tasks;
+      thread_cfgp->exp_thread_idx = tctr;
+
+      rc =
+         pthread_create(&new_thread, NULL, exec_one_of_n_threads, thread_cfgp);
+      ASSERT_EQUAL(0, rc);
+
+      thread_cfgp->this_thread_id = new_thread;
+   }
```
Again this should be using `platform_thread` and `task_thread_create`.

ajhconway commented on tests/unit/task_system_test.c:
```diff
@@ -0,0 +1,349 @@
+   // This is the important call to initialize thread-specific stuff in
+   // Splinter's task-system, which sets up the thread-id (index) and records
+   // this thread as active with the task system.
+   task_register_this_thread(thread_cfg->tasks, trunk_get_scratch_size());
```
Testing this is fine, but if we use the task thread creation machinery as I suggest above, then this call will happen as a hook.

This function is generally used only for the `main` thread or for threads otherwise created by a non-splinter context. So perhaps we should run this test on the main thread instead.

gapisback replied:
@ajhconway -- Not sure I understand what to change here.

With the revised set of tests, let me know if your comment above is still relevant and do I need to readjust the test code in any way

ajhconway replied:
Good now.

ajhconway reviewed (changes requested):
Thank you for doing this. These comments, tests and fixes are all very good.

A couple of testing changes would be good though.
1. The tests shouldn't call the pthread functions, but should call their platform and task wrappers (see comments in code).
2. We don't test `task_enqueue`, `task_perform_one` and `task_perform_all`, which is the primary way that the task system is used in the core code. It's fine if this is out of scope for this PR though.

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -101,6 +101,9 @@ platform_buffer_destroy(buffer_handle *bh)
 
+/*
+ * platform_thread_create() - External interface to create a Splinter thread.
+ */
```
(Added many such prologs to make the whole module easier to follow.)

gapisback commented on src/platform_linux/platform.c:
```diff
@@ -136,6 +139,12 @@ platform_thread_join(platform_thread thread)
 }
 
+platform_thread
+platform_thread_id_self()
```
Added, externalized for testing.

gapisback commented on src/task.c:
```diff
@@ -6,27 +6,31 @@
 #include "poison.h"
 
-#define MAX_HOOKS (2048)
+#define MAX_HOOKS (8)
```
Reduced this .... Is this good enough or too low?  Let's discuss.

gapisback commented on src/task.c:
```diff
@@ -40,8 +44,17 @@ task_init_tid_bitmask(uint64 *tid_bitmask)
+ */
 static void
-init_threadid(task_system *ts)
+task_init_threadid(task_system *ts)
```
Renamed with `task_` prefix to be consistent with other names in this file.

gapisback commented on src/task.c:
```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
+    * attempt to re-register a thread should find scratch space already
+    * allocated. Flag that.
+    */
+   platform_assert((thread_tid == 0)
```
Added ... Useful assertion. Found and tripped on incorrect usages in test code that was resulting in nested register thread for the same task.

gapisback commented on src/task.c:
```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
+   threadid tid = platform_get_tid();
+
+   // A registered thread should always have a non-zero tid (index).
+   if (tid == 0) {
```
Useful check ... Found bugs in test code that was doing 2 de-registers, inadvertently.

ajhconway replied:
Good catch

gapisback commented on src/task.c:
```diff
@@ -168,24 +266,43 @@ typedef struct {
-   thread_invoke         *thread_to_start = (thread_invoke *)func_and_args;
-   platform_thread_worker func            = thread_to_start->func;
-   void                  *arg             = thread_to_start->arg;
+   thread_invoke         *thread_started = (thread_invoke *)func_and_args;
```
Renamed variable from `thread_to_start` to `thread_started`, because by the time this function executes, it's already executing in the context of a newly fired up thread.

(Prior variable name was meaningful & correct in this fn's caller, where the thread was just going to be started.)

gapisback commented on src/task.h:
```diff
@@ -127,12 +127,25 @@ task_thread_create(const char            *name,
 
 // Register the calling thread, allocating scratch space for it
+#define task_register_this_thread(ts, scratch_size)                            \
+   task_register_thread((ts), (scratch_size), __FILE__, __LINE__, __FUNCTION__)
```
Changed this and on L141 below to become caller-macros, so I can pass-down code location tags to the lower level function, for improved diagnostics.

gapisback commented on tests/unit/task_system_test.c:
```diff
@@ -0,0 +1,496 @@
+ * done in its sibling test case, test_one_thread_using_lower_apis.
+ * ------------------------------------------------------------------------
+ */
+CTEST2(task_system, test_one_thread_using_extern_apis)
```
Added this new test case in 2nd commit, based on your review comments.

gapisback commented on tests/unit/task_system_test.c:
```diff
@@ -0,0 +1,496 @@
+ * checks accordingly.
+ */
+static void
+exec_one_thread_use_extern_apis(void *arg)
```
Worker fn added new in 2nd commit after your comments.

gapisback commented:
HI, @ajhconway -- I have revised the code and test code based on your feedback. Have done semi-major overhaul of the test code.

I think this should cover most of your feedback. The change set is back into the review cycle now

Thanks!

gapisback commented:
@ajhconway - I have addressed your comment (1), by reworking the test case and add a new one using just the external APIs. 

About (2): "_We don't test task_enqueue, task_perform_one and task_perform_all, which is the primary way that the task system is used in the core code. It's fine if this is out of scope for this PR though._"

I tried to work that in, but that will take me [more] time to understand how all these interfaces work and then to cobble together an execution flow which will invoke them. 

If you think that is very desirable, I can open an issue, but -- really -- I don't think I will have much time for that this month. (This work came out of Am behind my schedule to complete the PG-SplinterDB integration work, and am already running late on that.)

ajhconway reviewed (approved):
This looks good to me now, and I appreciate all the comments and cleanup.

One thing I would like is to preface the task tests with something to indicate that the external APIs are the preferred way of interacting and that the low level ones are not. I'm just trying to imagine a user looking at the test code as a template and trying to push them towards the external APIs.
